src/librustc_trans/mir/mod.rs

```diff
@@ -149,6 +149,15 @@ pub fn trans_mir<'blk, 'tcx: 'blk>(fcx: &'blk FunctionContext<'blk, 'tcx>) {
                                   TempRef::Lvalue(LvalueRef::alloca(&bcx,
                                                                     mty,
                                                                     &format!("temp{:?}", i)))
+                              } else if common::type_is_zero_size(bcx.ccx(), mty) {
+                                  // Zero-size temporaries aren't always initialized, which
+                                  // doesn't matter because they don't contain data, but
+                                  // we need something in the operand.
+                                  let op = OperandRef {
+                                      val: OperandValue::Immediate(common::C_nil(bcx.ccx())),
+                                      ty: mty
+                                  };
+                                  TempRef::Operand(Some(op))
```

Comment:
I feel like this condition ought to be abstracted into some `TempRef::new_operand` method.